src/goto-programs/remove_exceptions.cpp

```diff
@@ -98,175 +98,114 @@ class remove_exceptionst
+bool remove_exceptionst::may_throw_or_receive_exceptions(
+  const goto_programt &goto_program) const
+{
   forall_goto_program_instructions(instr_it, goto_program)
   {
-    bool has_uncaught_exceptions=false;
-    if(instr_it->is_function_call())
-    {
-      const exprt &function_expr=
-        to_code_function_call(instr_it->code).function();
-      DATA_INVARIANT(
-        function_expr.id()==ID_symbol,
-        "identifier expected to be a symbol");
-      const irep_idt &function_name=
-        to_symbol_expr(function_expr).get_identifier();
-      has_uncaught_exceptions=!exceptions_map[function_name].empty();
-    }
-
-    bool assertion_error=false;
     if(instr_it->is_throw())
     {
       const exprt &exc =
         uncaught_exceptions_domaint::get_exception_symbol(instr_it->code);
-      assertion_error =
+      bool is_assertion_error =
         id2string(uncaught_exceptions_domaint::get_exception_type(exc.type())).
         find("java.lang.AssertionError")!=std::string::npos;
+      if(!is_assertion_error)
+        return true;
     }
 
-    // if we find a throw different from AssertionError or a function call
-    // that may escape exceptions, then we add an exceptional return
-    // variable
-    if((instr_it->is_throw() && !assertion_error)
-       || has_uncaught_exceptions)
+    if(instr_it->is_function_call())
     {
-      // look up the function symbol
-      symbol_tablet::symbolst::const_iterator s_it=
-        symbol_table.symbols.find(function_id);
-
-      INVARIANT(
-        s_it!=symbol_table.symbols.end(),
-        "functions should be recorded in the symbol table");
-
-      auxiliary_symbolt new_symbol;
-      new_symbol.is_static_lifetime=true;
-      new_symbol.module=function_symbol.module;
-      new_symbol.base_name=id2string(function_symbol.base_name)+EXC_SUFFIX;
-      new_symbol.name=id2string(function_symbol.name)+EXC_SUFFIX;
-      new_symbol.mode=function_symbol.mode;
-      new_symbol.type=pointer_type(empty_typet());
-      symbol_table.add(new_symbol);
-
-      // initialize the exceptional return with NULL
-      symbol_exprt lhs_expr_null=new_symbol.symbol_expr();
-      null_pointer_exprt rhs_expr_null(pointer_type(empty_typet()));
-      goto_programt::targett t_null=
-        goto_program.insert_before(goto_program.instructions.begin());
-      t_null->make_assignment();
-      t_null->source_location=
-        goto_program.instructions.begin()->source_location;
-      t_null->code=code_assignt(
-        lhs_expr_null,
-        rhs_expr_null);
-      t_null->function=function_id;
-
-      break;
+      const exprt &function_expr=
+        to_code_function_call(instr_it->code).function();
+      DATA_INVARIANT(
+        function_expr.id()==ID_symbol,
+        "identifier expected to be a symbol");
+      const irep_idt &function_name=
+        to_symbol_expr(function_expr).get_identifier();
+      if(!exceptions_map[function_name].empty())
```

Comment:
Consider adding a comment explaining what a function having elements in the `exceptions_map` means?